### PR TITLE
test: add multiple LeafUpdater tests

### PR DESCRIPTION
**Pull request by gabriele-0201**

### CODE REVIEW

**Comment by gabriele-0201** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/thrumdev/nomt/643?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#643** <a href="https://app.graphite.dev/github/pr/thrumdev/nomt/643?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.dev/github/pr/thrumdev/nomt/643?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#642** <a href="https://app.graphite.dev/github/pr/thrumdev/nomt/642?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#641** <a href="https://app.graphite.dev/github/pr/thrumdev/nomt/641?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `master`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

File: nomt/src/beatree/ops/update/leaf_updater.rs [hunk 1]
```diff
@@ -903,4 +909,265 @@ mod tests {
...
+        let leaf_1 = &new_leaves.inner.get(&[0; 32]).unwrap().0;
+        assert_eq!(leaf_1.n(), 3);
+        let leaf_body_size = body_size(leaf_1.n(), leaf_1.values_size(0, leaf_1.n()));
+        assert!(leaf_body_size < midpoint);
+        let leaf_2 = &new_leaves.inner.get(&separate(&key(3), &key(4))).unwrap().0;
+        assert_eq!(leaf_2.n(), 3);
+    }
+
+    #[test]
+    fn split_left_node_impossible_not_halfull() {
```
Comment (rphmeier) [inline nomt/src/beatree/ops/update/leaf_updater.rs:939]: half_full

(I've also learned the term "rightsized" to refer to nodes that are between the minimum and maximum)

File: nomt/src/beatree/ops/update/leaf_updater.rs [hunk 2]
```diff
@@ -903,4 +909,265 @@ mod tests {
...
+        let mut new_leaves = TestHandleNewLeaf::default();
+
+        updater.ingest(key(1), Some(vec![1; 1300]), false, |_| {});
+        updater.ingest(key(2), Some(vec![1; 650]), false, |_| {});
+        updater.ingest(key(3), Some(vec![1; MAX_LEAF_VALUE_SIZE]), false, |_| {});
+        updater.ingest(key(4), Some(vec![1; 678]), false, |_| {});
+
+        // The first two sum up to a body_size of 2018 which is less than
+        // LEAF_MERGE_THRESHOLD but even followd by the biggest leaf value possible
+        // it's not possible to end up in an  to overfull scenario on the first
```
Comment (rphmeier) [inline nomt/src/beatree/ops/update/leaf_updater.rs:950]: right, this is what I mentioned in the comment on the last PR. The `MAX_LEAF_VALUE_SIZE` is chosen for that reason.

File: nomt/src/beatree/ops/update/leaf_updater.rs (hunk 1 above)
Comment (gabriele-0201) [reply nomt/src/beatree/ops/update/leaf_updater.rs:939]: Thank you! `split_left_node_always_rightsized` is perfect!

Not sure why in this stack I was convinced that half-full is written as halfull and also other words around a similar error :sweat_smile:

**Comment by rphmeier** [comment]
### Merge activity

* **Dec 20, 12:11 PM EST**: A user started a stack merge that includes this pull request via [Graphite](https://app.graphite.dev/github/pr/thrumdev/nomt/643).

**Pull request merged**